### PR TITLE
[Ready] Supercedes #561 - Fix Docker Action SDK

### PR DESCRIPTION
**Pull request by rabbah**
Supercedes #561.

Fix all parts of the docker skeleton action which were completely broken.

- Check `clientApp` returned a JSON parsable object and if not, report error and log a message.
- Update example to show how params are received and return a proper JSON.
- Fix space in path in various build scripts.
- Add gradle build for docker skeleton - to enable testing locally.
- Replace broken docker action tests with test that verifies the docker skeleton can tolerate an init message and can receive/respond with JSON objects. This test suite is not complete in that the docker skeleton hardcodes an example and the tests are specific to that example so cannot test that failure modes are correct yet.

### CODE REVIEW

**Comment by rabbah** [comment]
@csantanapr FYI you may want to review the readme steps.

**Comment by rabbah** [comment]
@domdom82 can you exclude gradle file from the docker skeleton?

**Comment by rabbah** [comment]
PG 381.

File: sdk/docker/dockerSkeleton/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+FROM buildpack-deps:trusty-curl
+
+ENV DEBIAN_FRONTEND noninteractive
+
+# Initial update and some basics.
+# This odd double update seems necessary to get curl to download without 404 errors.
+RUN apt-get update --fix-missing && \
+apt-get install -y wget && \
+apt-get update && \
+apt-get install -y curl && \
```
Comment (rabbah) [inline sdk/docker/dockerSkeleton/Dockerfile]: These can go too

File: sdk/docker/dockerSkeleton/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+apt-get update
+
+# Get gcc
+RUN apt-get update --fix-missing && \
+apt-get -y install gcc
+
+# install nodejs and npm
+RUN curl -sL https://deb.nodesource.com/setup_0.12 | bash - && \
```
Comment (markusthoemmes) [inline sdk/docker/dockerSkeleton/Dockerfile]: Might want to change this to the alternative "just download the binary" approach. Smaller image, less dependencies.

File: tools/cli/build.gradle [hunk 3]
```diff
@@ -38,8 +38,7 @@ distDocker.dependsOn(establishStructure)
 task distTar(dependsOn: [removeDistribution, distDocker]) << {
     run("docker rm -f cli", true) // Ignore error as this is a cleanup measure
     run("docker run --name cli whisk/cli")
-    run("docker cp cli:/cli/dist/ ${buildDir}/distributions/")
-    run("docker rm -f cli")
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: Why remove this? Need to clean up after the CLI was extracted from the container?

Comment (rabbah) [inline tools/cli/build.gradle]: bad merge - thanks for catching.

File: sdk/docker/dockerSkeleton/Dockerfile (hunk 2 above)
Comment (rabbah) [inline sdk/docker/dockerSkeleton/Dockerfile]: K - do we do that for the nodejs image?

Comment (rabbah) [inline sdk/docker/dockerSkeleton/Dockerfile]: Yup we do.

Comment (markusthoemmes) [inline sdk/docker/dockerSkeleton/Dockerfile]: These two lines need to go away then.

**Comment by rabbah** [comment]
@markusthoemmes fixed.

**Comment by markusthoemmes** [comment]
Do we need the extra subfolder `dockerSkeleton`?

**Comment by csantanapr** [comment]
I agree with @markusthoemmes let's not have an extra folder `dockerSkeleton` source control (github).
I prefer to have it more flat, and aligns better if we move the sdk to it's own repo (i.e. openwhisk-sdk-docker)

I'm ok if the sdk-docker.tar.gz contains an extra folder inside i.e. `openwhisk-sdk-docker`

File: tools/build/scanCode.py [hunk 4]
```diff
@@ -1,3 +1,21 @@
+#!/usr/bin/env python
```
Comment (psuter) [inline tools/build/scanCode.py]: I believe you have these changes in two distinct PRs. Should disappear when rebasing.

File: sdk/docker/client/clientApp [hunk 5]
```diff
@@ -5,4 +5,4 @@
 # in this folder.  For more info, see the Readme.md in 
 # parent directory.
 #
-echo "Hello $1 from custom binary"
+echo '{ "msg": "Hello from arbitrary C program!" }'
```
Comment (psuter) [inline sdk/docker/client/clientApp]: Why does the stub bash script claim it's a C program?

File: tests/src/actionContainers/DockerSkeletonActionContainerTests.scala [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        val (out, err) = withDockerSdkContainer { c =>
+            val args = List(
+                JsObject(),
+                JsObject("numbers" -> JsArray(JsNumber(42), JsNumber(1))))
+
+            for (arg <- args) {
+                val (runCode, out) = c.run(runPayload(arg))
+
+                runCode should be(200)
+                out shouldBe a[Some[_]]
```
Comment (psuter) [inline tests/src/actionContainers/DockerSkeletonActionContainerTests.scala]: `shouldBe defined`

File: sdk/docker/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# build/push, argument is your docker hub user name and a valid docker image name
+> ./buildAndPush <dockerhub username>/whiskexample
+# create docker action
+> wsk action create dockerSkeletonExample --docker <dockerhub username>/whiskExample
+# invoke created action
+> wsk action invoke dockerSkeletonExample --blocking
+```
+
+The binary to execute must be located in the `client` folder. The name of the binary
+must be `clientApp`.
```
Comment (sjfink) [inline sdk/docker/README.md]: I've never liked the name client/clientApp.

How about naming the action binary 'action'?

**Comment by psuter** [comment]
Tests look good to me. Will wait for PG build. We should as usual be careful and make sure the Docker image is available on all testing machines if running a distributed build.

File: sdk/docker/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+6. Test out action with CLI
+
+The script `buildAndPush.sh` is provided for your convenience. The following command sequence
+runs the included example Docker action container using OpenWhisk.
+
+```
+# install ./dockerSkeleton with example
+> wsk sdk install docker
+# change working directory
+> cd dockerSkeleton
```
Comment (sjfink) [inline sdk/docker/README.md]: In latest zip I got from @rabbah dockerSkeleton is renamed 'docker'

Comment (rabbah) [inline sdk/docker/README.md]: @sjfink I am in the process of moving per suggestions above. You got an early preview. 

Will rename clientApp.

**Comment by rabbah** [comment]
Addressed comments in https://github.com/openwhisk/openwhisk/pull/652/commits/296d17fd061a4fc1416804e95bc4225b0cffe895.

PG pending.

**Comment by rabbah** [comment]
PG 99.

**Comment by rabbah** [comment]
PG 104 approved.

**Pull request merged**